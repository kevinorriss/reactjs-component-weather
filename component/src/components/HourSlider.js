import React from 'react'
import PropTypes from 'prop-types'
import Slider from 'rc-slider'
import axios from 'axios'
import moment from 'moment-timezone'
import 'rc-slider/assets/index.css'

class HourSlider extends React.Component {
    constructor(props) {
        super(props)

        // bind this to the functions
        this.getForecast = this.getForecast.bind(this)

        // set the initial state
        this.state = {
            markCount: 12,
            markStep: 2,
            hourStep: 2,
            marks: {},
            disabled: true
        }
    }

    componentDidUpdate(prevProps) {
        // if the lat/long was null and has been updated, we have the location
        if (!prevProps.latitude && !prevProps.longitude && this.props.latitude && this.props.longitude) {
            this.getForecast(this.props.latitude, this.props.longitude)
        }
    }

    async requestForecast(url) {
        return await axios.get(url)
    }

    async getForecast(latitude, longitude) {
        let response
        try {
            // call the REST API to get the forecast data for the provided location
            const url = `${this.props.forecastURL}?latitude=${latitude}&longitude=${longitude}`
            response = await this.requestForecast(url)
        } catch (error) {
            // fire the error callback function and return
            this.props.onForecastError('Error getting forecast response')
            return
        }

        // destructure the forecast
        const { timezone, hourly } = response.data

        const maxMark = (this.state.markCount - 1) * this.state.hourStep

        // check that we have enough hourly forecasts for each step
        if (hourly.length < maxMark) {
            this.props.onForecastError("Hourly forecast not covered")
            return
        }

        // fire the forecast callback function
        this.props.onForecastReceived(response.data)

        let marks = {}
        let count = 0
        for (let i = 0; i <= maxMark; i += this.state.hourStep) {
            if (count % this.state.markStep == 0) {
                marks[count] = moment.unix(hourly[i].time).tz(timezone).format('HH:mm')
            } else {
                marks[count] = ''
            }
            count++;
        }

        this.setState((prevState) => ({
            ...prevState,
            marks,
            disabled: false
        }))
    }

    render() {
        return (
            <Slider marks={this.state.marks}
                step={1}
                max={this.state.markCount-1}
                included={false}
                disabled={this.state.disabled}
                onChange={(index) => { this.props.onSliderChange(index*this.state.hourStep) }}
            />
        )
    }
}

HourSlider.propTypes = {
    latitude: PropTypes.number,
    longitude: PropTypes.number,
    onForecastReceived: PropTypes.func.isRequired,
    onForecastError: PropTypes.func.isRequired,
    onSliderChange: PropTypes.func.isRequired,
    forecastURL: PropTypes.string.isRequired,
}

export default HourSlider
