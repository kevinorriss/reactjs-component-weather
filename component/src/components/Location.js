import React from 'react'
import PropTypes from 'prop-types'
const axios = require('axios')

class Location extends React.Component {
    constructor(props) {
        super(props)
        
        // bind this to the functions
        this.onPositionSuccess = this.onPositionSuccess.bind(this)

        // set the initial state
        this.state = {
            name: 'Waiting for location'
        }
    }

    componentDidMount() {
        // if browser doesn't support geolocation, fire event and return
        if (typeof navigator.geolocation === "undefined") {
            // set the error text
            this.setState({ name: Location.ERROR_TEXT })

            // call the location error callback
            this.props.onLocationError(Location.UNSUPPORTED_BROWSER)

            return
        }

        // request the location from the browser
        navigator.geolocation.getCurrentPosition(
            // success
            this.onPositionSuccess,
            // error
            ({ code }) => {
                // pass the error to the callback
                this.props.onLocationError(code)

                // set the name in the state
                this.setState({ name: Location.ERROR_TEXT})
            }
        )
    }

    async onPositionSuccess(position) {
        try {
            // destructure the position coordinates to lat/long
            const { latitude, longitude } = position.coords

            // fire callback function
            this.props.onLocationReceived(latitude, longitude)

            // call the REST API to get the location name
            const url = `${this.props.locationURL}?latitude=${latitude}&longitude=${longitude}`
            const response = await axios.get(url)

            // set the name in the state
            this.setState({ name: response.data })

            // call the name received callback
            if (this.props.onNameReceived) {
                this.props.onNameReceived(response.data)
            }
        } catch (error) {
            // set the location name
            this.setState({ name: Location.ERROR_TEXT})

            // fire error callback function
            this.props.onLocationError(Location.RESPONSE_ERROR)
        }
    }

    render() {
        return (
            <div className="weather__location">{this.state.name}</div>
        )
    }
}

Location.PERMISSION_DENIED = 1
Location.POSITION_UNAVAILABLE = 2
Location.TIMEOUT = 3
Location.UNSUPPORTED_BROWSER = 4
Location.RESPONSE_ERROR = 5
Location.ERROR_TEXT = 'Location unavailable...'

Location.propTypes = {
    onLocationReceived: PropTypes.func.isRequired,
    onLocationError: PropTypes.func.isRequired,
    onNameReceived: PropTypes.func,
    locationURL: PropTypes.string.isRequired
}

export default Location