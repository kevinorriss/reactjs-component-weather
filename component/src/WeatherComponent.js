import React from 'react'
import PropTypes from 'prop-types'
import Moment from 'react-moment'
import Location from './components/Location'
import HourSlider from './components/HourSlider'
import Icon from './components/Icon'
import './styles.css'

class WeatherComponent extends React.Component {
    constructor(props) {
        super(props)
        this.onLocationReceived = this.onLocationReceived.bind(this)
        this.onLocationError = this.onLocationError.bind(this)
        this.onForecastReceived = this.onForecastReceived.bind(this)
        this.onForecastError = this.onForecastError.bind(this)
        this.onSliderChange = this.onSliderChange.bind(this)

        this.state = {
            position: {
                latitude: null,
                longitude: null
            },
            timezone: null,
            daily: null,
            hourly: null,
            currentHour: null,
            error: null
        }
    }

    onLocationReceived(latitude, longitude) {
        // set the position in the state
        this.setState((prevState) => ({
            ...prevState,
            position: {
                latitude,
                longitude
            }
        }))
    }

    onLocationError(code) {
        // pick the error message depending on code passed
        let errorMessage
        switch (code) {
            case Location.PERMISSION_DENIED:
                errorMessage = 'Permission to read your location was denied'
                break
            case Location.POSITION_UNAVAILABLE:
                errorMessage = 'Your position is unavailable'
                break
            case Location.TIMEOUT:
                errorMessage = 'Getting your position timed out'
                break
            case Location.UNSUPPORTED_BROWSER:
                errorMessage = 'Your browser does not support geolocation'
                break
            case Location.RESPONSE_ERROR:
                errorMessage = 'An error occurred in the response to getting your location information'
                break
            default:
                errorMessage = 'An unknown error occured'
                break
        }

        // set the error in the state
        this.setState((prevState) => ({
            ...prevState,
            error: errorMessage
        }))
    }

    onNameReceived(name) {
        console.log(`onNameReceived:${name}`)
    }

    onForecastReceived(forecast) {
        this.setState((prevState) => ({
            ...prevState,
            timezone: forecast.timezone,
            daily: forecast.daily,
            hourly: forecast.hourly,
            currentHour: forecast.hourly[0]
        }))
    }

    onForecastError() {
        console.log('forecast error')
    }

    onSliderChange(index) {
        this.setState((prevState) => ({
            ...prevState,
            currentHour: prevState.hourly[index]
        }))
    }

    render() {
        return (
            <div className="weather__container">
                {this.state.error && (
                    <div className="weather__error">
                        <p>{this.state.error}</p>
                    </div>
                )}
                <Location
                    onLocationReceived={this.onLocationReceived}
                    onNameReceived={this.onNameReceived}
                    onLocationError={this.onLocationError}
                    locationURL={this.props.locationURL}/>
                <div>
                    <p className="weather__current-summary" 
                        title={this.state.currentHour && this.state.currentHour.summary}>
                        {this.state.currentHour ? (
                            <React.Fragment>
                                <Moment unix
                                    format="ddd HH:mm"
                                    tz={this.state.timezone}>
                                    {this.state.currentHour.time}
                                </Moment>, {this.state.currentHour.summary}
                            </React.Fragment>
                        ) : (
                            <span>Loading weather data...</span>
                        )
                    }</p>
                    <div className="weather__current-forecast">
                        {this.state.currentHour ? (
                            <React.Fragment>
                                <Icon name={this.state.currentHour.icon}/>
                                <div className="weather__current-temp">
                                    {Math.floor(this.state.currentHour.temperature)}&deg;C
                                </div>
                                <div className="weather__current-perc">
                                    <p>Precip: {Math.floor(this.state.currentHour.precipProbability*100)}%</p>
                                    <p>Humidity: {Math.floor(this.state.currentHour.humidity*100)}%</p>
                                    <p>Wind: {Math.floor(this.state.currentHour.windSpeed)} mph</p>
                                </div>
                            </React.Fragment>
                        ) : (
                            <React.Fragment>
                                <Icon name="default" />
                                <div className="weather__current-temp">--</div>
                                <div className="weather__current-perc">
                                    <p>Precip: --</p>
                                    <p>Humidity: --</p>
                                    <p>Wind: --</p>
                                </div>
                            </React.Fragment>
                        )}
                    </div>
                </div>
                <div className="weather__slider">
                <HourSlider
                    latitude={this.state.position.latitude}
                    longitude={this.state.position.longitude}
                    onForecastReceived={this.onForecastReceived}
                    onForecastError={this.onForecastError}
                    forecastURL={this.props.forecastURL}
                    onSliderChange={this.onSliderChange}/>
                </div>
                <div className="weather__daily">
                {this.state.daily ? (
                    this.state.daily.map((day, index) => (
                        <div className="weather__day" title={day.summary} key={index}>
                            <Moment unix format="ddd" tz={this.state.timezone}>{day.time}</Moment>
                            <Icon name={day.icon} />
                            <p className="weather__temp">{Math.floor(day.temperatureMax)}&deg;C</p>
                            <p className="weather__temp weather__temp--low">{Math.floor(day.temperatureMin)}&deg;C</p>
                        </div>
                    ))
                ) : (
                    <div className="weather__day">
                        <Moment unix format="[--]">{0}</Moment>
                        <Icon name="default" />
                        <p className="weather__temp">--</p>
                        <p className="weather__temp weather__temp--low">--</p>
                    </div>
                )}
                </div>
                <p className="weather__icon-link">
                    <a href="https://www.freepik.com/free-vector/weather-icons-set_709126.htm" 
                        target="_blank">Icons created by freepik - www.freepik.com</a>
                </p>
            </div>
        )
    }
}

WeatherComponent.propTypes = {
    locationURL: PropTypes.string.isRequired,
    forecastURL: PropTypes.string.isRequired
}

export default WeatherComponent