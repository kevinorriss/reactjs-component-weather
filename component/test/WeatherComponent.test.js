import WeatherComponent from '../src/WeatherComponent'
import Location from '../src/components/Location'
import HourSlider from '../src/components/HourSlider'

let positionOne
let mockName
let forecast

const onLocationReceived = jest.spyOn(WeatherComponent.prototype, 'onLocationReceived')
const onLocationError = jest.spyOn(WeatherComponent.prototype, 'onLocationError')
const onForecastReceived = jest.spyOn(WeatherComponent.prototype, 'onForecastReceived')
const onForecastError = jest.spyOn(WeatherComponent.prototype, 'onForecastError')
const onSliderChange = jest.spyOn(WeatherComponent.prototype, 'onSliderChange')

beforeEach(() => {
    // test location
    positionOne = {
        coords: {
            latitude: 51.505455,
            longitude: -0.075356
        }
    }

    mockName = 'London, England'

    forecast = {
        timezone: 'Europe/London',
        hourly: [
            {time: 1582848000, summary: 'Rain', icon: 'rain', precipProbability: 0.85, temperature: 7.34, humidity: 0.82, windSpeed: 6.31},
            {time: 1582851600, summary: 'Rain', icon: 'rain', precipProbability: 0.82, temperature: 7.34, humidity: 0.82, windSpeed: 6.31},
            {time: 1582855200, summary: 'Rain', icon: 'rain', precipProbability: 0.82, temperature: 7.34, humidity: 0.82, windSpeed: 6.31},
            {time: 1582858800, summary: 'Rain', icon: 'rain', precipProbability: 0.82, temperature: 7.34, humidity: 0.82, windSpeed: 6.31},
            {time: 1582862400, summary: 'Rain', icon: 'rain', precipProbability: 0.82, temperature: 7.34, humidity: 0.82, windSpeed: 6.31},
            {time: 1582866000, summary: 'Rain', icon: 'rain', precipProbability: 0.82, temperature: 7.34, humidity: 0.82, windSpeed: 6.31},
            {time: 1582869600, summary: 'Rain', icon: 'rain', precipProbability: 0.82, temperature: 7.34, humidity: 0.82, windSpeed: 6.31},
            {time: 1582873200, summary: 'Rain', icon: 'rain', precipProbability: 0.82, temperature: 7.34, humidity: 0.82, windSpeed: 6.31},
            {time: 1582876800, summary: 'Rain', icon: 'rain', precipProbability: 0.82, temperature: 7.34, humidity: 0.82, windSpeed: 6.31},
            {time: 1582880400, summary: 'Rain', icon: 'rain', precipProbability: 0.82, temperature: 7.34, humidity: 0.82, windSpeed: 6.31},
            {time: 1582884000, summary: 'Rain', icon: 'rain', precipProbability: 0.82, temperature: 7.34, humidity: 0.82, windSpeed: 6.31},
            {time: 1582887600, summary: 'Rain', icon: 'rain', precipProbability: 0.82, temperature: 7.34, humidity: 0.82, windSpeed: 6.31},
            {time: 1582891200, summary: 'Rain', icon: 'rain', precipProbability: 0.82, temperature: 7.34, humidity: 0.82, windSpeed: 6.31},
            {time: 1582894800, summary: 'Rain', icon: 'rain', precipProbability: 0.82, temperature: 7.34, humidity: 0.82, windSpeed: 6.31},
            {time: 1582898400, summary: 'Rain', icon: 'rain', precipProbability: 0.82, temperature: 7.34, humidity: 0.82, windSpeed: 6.31},
            {time: 1582902000, summary: 'Rain', icon: 'rain', precipProbability: 0.82, temperature: 7.34, humidity: 0.82, windSpeed: 6.31},
            {time: 1582905600, summary: 'Rain', icon: 'rain', precipProbability: 0.82, temperature: 7.34, humidity: 0.82, windSpeed: 6.31},
            {time: 1582909200, summary: 'Rain', icon: 'rain', precipProbability: 0.82, temperature: 7.34, humidity: 0.82, windSpeed: 6.31},
            {time: 1582912800, summary: 'Rain', icon: 'rain', precipProbability: 0.82, temperature: 7.34, humidity: 0.82, windSpeed: 6.31},
            {time: 1582916400, summary: 'Rain', icon: 'rain', precipProbability: 0.82, temperature: 7.34, humidity: 0.82, windSpeed: 6.31},
            {time: 1582920000, summary: 'Rain', icon: 'rain', precipProbability: 0.82, temperature: 7.34, humidity: 0.82, windSpeed: 6.31},
            {time: 1582923600, summary: 'Rain', icon: 'rain', precipProbability: 0.82, temperature: 7.34, humidity: 0.82, windSpeed: 6.31},
        ],
        daily: [
            {time: 1582416000, summary: 'Light rain', icon: 'rain', temperatureMin: 6.37, temperatureMax: 13.47},
            {time: 1582516000, summary: 'Overcast', icon: 'cloudy', temperatureMin: 4.23, temperatureMax: 12.42}
        ]
    }

    // mock getting the position from the browser
    navigator.geolocation = {
        getCurrentPosition: (success, error) => {
            success(positionOne)
        }
    }

    const locationRequest = jest.spyOn(Location.prototype, 'requestLocation')
    locationRequest.mockImplementationOnce(() => Promise.resolve({ data: mockName}))

    const forecastRequest = jest.spyOn(HourSlider.prototype, 'requestForecast')
    forecastRequest.mockImplementationOnce(() => Promise.resolve({ data: forecast}))
})

afterEach(() => {
    onLocationReceived.mockClear()
    onLocationError.mockClear()
    onForecastReceived.mockClear()
    onForecastError.mockClear()
    onSliderChange.mockClear()
})

it('Should render initial state correctly', (done) => {
    // mock the location received callback to do nothing
    onLocationReceived.mockImplementationOnce((latitude, longitude) => {})

    // mount the component
    const wrapper = mount(
        <WeatherComponent
            locationURL="/weather/location"
            forecastURL="/weather/forecast" />
    )

    // set a short delay before testing callbacks
    setTimeout(() => {
        // the mocked location received callback should have been called
        expect(onLocationReceived).toHaveBeenCalledTimes(1)

        // the state should have remained the same
        expect(wrapper.state()).toEqual({
            position: {
                latitude: null,
                longitude: null
            },
            timezone: null,
            daily: null,
            hourly: null,
            currentHour: null,
            error: null
        })

        // component should render as expected
        expect(wrapper).toMatchSnapshot()
        
        // tell jest the test is complete
        done()
    }, 20)
})

it('Should handle location received in callback', (done) => {
    // mount the component
    const wrapper = mount(
        <WeatherComponent
            locationURL="/weather/location"
            forecastURL="/weather/forecast" />
    )

    // set a short delay before testing callbacks
    setTimeout(() => {
        // the mocked location received callback should have been called
        expect(onLocationReceived).toHaveBeenCalledTimes(1)
        expect(onLocationReceived).toHaveBeenCalledWith(positionOne.coords.latitude, positionOne.coords.longitude)

        // the state should have remained the same
        expect(wrapper.state('position')).toEqual(positionOne.coords)

        // component should render as expected
        expect(wrapper).toMatchSnapshot()
        
        // tell jest the test is complete
        done()
    }, 20)
})