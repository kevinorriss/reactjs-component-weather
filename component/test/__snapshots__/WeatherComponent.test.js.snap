// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should handle location received in callback 1`] = `
<WeatherComponent
  forecastURL="/weather/forecast"
  locationURL="/weather/location"
>
  <div
    className="weather__container"
  >
    <Location
      locationURL="/weather/location"
      onLocationError={[Function]}
      onLocationReceived={[Function]}
      onNameReceived={[Function]}
    >
      <div
        className="weather__location"
      >
        Waiting for location
      </div>
    </Location>
    <div>
      <p
        className="weather__current-summary"
        title={null}
      >
        <span>
          Loading weather data...
        </span>
      </p>
      <div
        className="weather__current-forecast"
      >
        <HourSlider
          name="default"
        >
          <img
            className="weather__icon"
            src="<svg version=\\"1.1\\" id=\\"Layer_1\\" xmlns=\\"http://www.w3.org/2000/svg\\" x=\\"0\\" y=\\"0\\" width=\\"100\\" height=\\"100\\" xml:space=\\"preserve\\"><path d=\\"M63.295 43.878h-.125l-.525 2.774-1.274 3.873h3.748l-1.301-3.898z\\"/><path d=\\"M50 20c-16.568 0-30 13.432-30 30s13.432 30 30 30c16.569 0 30-13.432 30-30S66.568 20 50 20zm-7.448 37.273h-2.499l-6.572-9.247-1.375-2.75h-.125l.399 2.75v9.122h-3.248V39.53h2.499l6.622 9.321 1.325 2.674h.125l-.4-2.674v-9.196h3.249v17.618zm4.848 3.373l-2.449-1.149 8.921-20.143 2.449 1.1L47.4 60.646zm19.868-3.498l-1.398-3.873h-5.373l-1.324 3.873h-3.424l6.297-17.62h2.523l6.299 17.62h-3.6z\\"/></svg>"
          />
        </HourSlider>
        <div
          className="weather__current-temp"
        >
          --
        </div>
        <div
          className="weather__current-perc"
        >
          <p>
            Precip: --
          </p>
          <p>
            Humidity: --
          </p>
          <p>
            Wind: --
          </p>
        </div>
      </div>
    </div>
    <div
      className="weather__slider"
    >
      <HourSlider
        forecastURL="/weather/forecast"
        latitude={51.505455}
        longitude={-0.075356}
        onForecastError={[Function]}
        onForecastReceived={[Function]}
        onSliderChange={[Function]}
      >
        <ComponentEnhancer(undefined)
          activeDotStyle={Object {}}
          className=""
          disabled={true}
          dotStyle={Object {}}
          dots={false}
          handle={[Function]}
          handleStyle={
            Array [
              Object {},
            ]
          }
          included={false}
          marks={Object {}}
          max={11}
          min={0}
          onAfterChange={[Function]}
          onBeforeChange={[Function]}
          onChange={[Function]}
          prefixCls="rc-slider"
          railStyle={Object {}}
          reverse={false}
          step={1}
          trackStyle={
            Array [
              Object {},
            ]
          }
          vertical={false}
        >
          <div
            className="rc-slider rc-slider-disabled"
            onBlur={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onMouseDown={[Function]}
            onMouseUp={[Function]}
            onTouchStart={[Function]}
          >
            <div
              className="rc-slider-rail"
              style={Object {}}
            />
            <Track
              className="rc-slider-track"
              included={false}
              length={0}
              offset={0}
              reverse={false}
              style={Object {}}
              vertical={false}
            />
            <Steps
              activeDotStyle={Object {}}
              dotStyle={Object {}}
              dots={false}
              included={false}
              lowerBound={0}
              marks={Object {}}
              max={11}
              min={0}
              prefixCls="rc-slider"
              reverse={false}
              step={1}
              upperBound={0}
              vertical={false}
            >
              <div
                className="rc-slider-step"
              />
            </Steps>
            <Handle
              className="rc-slider-handle"
              disabled={true}
              key="0"
              max={11}
              min={0}
              offset={0}
              prefixCls="rc-slider"
              reverse={false}
              style={Object {}}
              value={0}
              vertical={false}
            >
              <div
                aria-disabled={true}
                aria-valuemax={11}
                aria-valuemin={0}
                aria-valuenow={0}
                className="rc-slider-handle"
                onBlur={[Function]}
                onKeyDown={[Function]}
                onMouseDown={[Function]}
                role="slider"
                style={
                  Object {
                    "left": "0%",
                    "right": "auto",
                    "transform": "translateX(-50%)",
                  }
                }
                tabIndex={null}
              />
            </Handle>
            <Marks
              className="rc-slider-mark"
              included={false}
              lowerBound={0}
              marks={Object {}}
              max={11}
              min={0}
              onClickLabel={[Function]}
              reverse={false}
              upperBound={0}
              vertical={false}
            >
              <div
                className="rc-slider-mark"
              />
            </Marks>
          </div>
        </ComponentEnhancer(undefined)>
      </HourSlider>
    </div>
    <div
      className="weather__daily"
    >
      <div
        className="weather__day"
      >
        <t
          ago={false}
          calendar={false}
          decimal={false}
          element={null}
          filter={[Function]}
          format="[--]"
          fromNow={false}
          interval={60000}
          local={false}
          onChange={[Function]}
          titleFormat=""
          toNow={false}
          unit={null}
          unix={true}
          utc={false}
          withTitle={false}
        >
          <time
            dateTime={"1970-01-01T00:00:00.000Z"}
          >
            --
          </time>
        </t>
        <HourSlider
          name="default"
        >
          <img
            className="weather__icon"
            src="<svg version=\\"1.1\\" id=\\"Layer_1\\" xmlns=\\"http://www.w3.org/2000/svg\\" x=\\"0\\" y=\\"0\\" width=\\"100\\" height=\\"100\\" xml:space=\\"preserve\\"><path d=\\"M63.295 43.878h-.125l-.525 2.774-1.274 3.873h3.748l-1.301-3.898z\\"/><path d=\\"M50 20c-16.568 0-30 13.432-30 30s13.432 30 30 30c16.569 0 30-13.432 30-30S66.568 20 50 20zm-7.448 37.273h-2.499l-6.572-9.247-1.375-2.75h-.125l.399 2.75v9.122h-3.248V39.53h2.499l6.622 9.321 1.325 2.674h.125l-.4-2.674v-9.196h3.249v17.618zm4.848 3.373l-2.449-1.149 8.921-20.143 2.449 1.1L47.4 60.646zm19.868-3.498l-1.398-3.873h-5.373l-1.324 3.873h-3.424l6.297-17.62h2.523l6.299 17.62h-3.6z\\"/></svg>"
          />
        </HourSlider>
        <p
          className="weather__temp"
        >
          --
        </p>
        <p
          className="weather__temp weather__temp--low"
        >
          --
        </p>
      </div>
    </div>
    <p
      className="weather__icon-link"
    >
      <a
        href="https://www.freepik.com/free-vector/weather-icons-set_709126.htm"
        target="_blank"
      >
        Icons created by freepik - www.freepik.com
      </a>
    </p>
  </div>
</WeatherComponent>
`;

exports[`Should render initial state correctly 1`] = `
<WeatherComponent
  forecastURL="/weather/forecast"
  locationURL="/weather/location"
>
  <div
    className="weather__container"
  >
    <Location
      locationURL="/weather/location"
      onLocationError={[Function]}
      onLocationReceived={[Function]}
      onNameReceived={[Function]}
    >
      <div
        className="weather__location"
      >
        Waiting for location
      </div>
    </Location>
    <div>
      <p
        className="weather__current-summary"
        title={null}
      >
        <span>
          Loading weather data...
        </span>
      </p>
      <div
        className="weather__current-forecast"
      >
        <HourSlider
          name="default"
        >
          <img
            className="weather__icon"
            src="<svg version=\\"1.1\\" id=\\"Layer_1\\" xmlns=\\"http://www.w3.org/2000/svg\\" x=\\"0\\" y=\\"0\\" width=\\"100\\" height=\\"100\\" xml:space=\\"preserve\\"><path d=\\"M63.295 43.878h-.125l-.525 2.774-1.274 3.873h3.748l-1.301-3.898z\\"/><path d=\\"M50 20c-16.568 0-30 13.432-30 30s13.432 30 30 30c16.569 0 30-13.432 30-30S66.568 20 50 20zm-7.448 37.273h-2.499l-6.572-9.247-1.375-2.75h-.125l.399 2.75v9.122h-3.248V39.53h2.499l6.622 9.321 1.325 2.674h.125l-.4-2.674v-9.196h3.249v17.618zm4.848 3.373l-2.449-1.149 8.921-20.143 2.449 1.1L47.4 60.646zm19.868-3.498l-1.398-3.873h-5.373l-1.324 3.873h-3.424l6.297-17.62h2.523l6.299 17.62h-3.6z\\"/></svg>"
          />
        </HourSlider>
        <div
          className="weather__current-temp"
        >
          --
        </div>
        <div
          className="weather__current-perc"
        >
          <p>
            Precip: --
          </p>
          <p>
            Humidity: --
          </p>
          <p>
            Wind: --
          </p>
        </div>
      </div>
    </div>
    <div
      className="weather__slider"
    >
      <HourSlider
        forecastURL="/weather/forecast"
        latitude={null}
        longitude={null}
        onForecastError={[Function]}
        onForecastReceived={[Function]}
        onSliderChange={[Function]}
      >
        <ComponentEnhancer(undefined)
          activeDotStyle={Object {}}
          className=""
          disabled={true}
          dotStyle={Object {}}
          dots={false}
          handle={[Function]}
          handleStyle={
            Array [
              Object {},
            ]
          }
          included={false}
          marks={Object {}}
          max={11}
          min={0}
          onAfterChange={[Function]}
          onBeforeChange={[Function]}
          onChange={[Function]}
          prefixCls="rc-slider"
          railStyle={Object {}}
          reverse={false}
          step={1}
          trackStyle={
            Array [
              Object {},
            ]
          }
          vertical={false}
        >
          <div
            className="rc-slider rc-slider-disabled"
            onBlur={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onMouseDown={[Function]}
            onMouseUp={[Function]}
            onTouchStart={[Function]}
          >
            <div
              className="rc-slider-rail"
              style={Object {}}
            />
            <Track
              className="rc-slider-track"
              included={false}
              length={0}
              offset={0}
              reverse={false}
              style={Object {}}
              vertical={false}
            />
            <Steps
              activeDotStyle={Object {}}
              dotStyle={Object {}}
              dots={false}
              included={false}
              lowerBound={0}
              marks={Object {}}
              max={11}
              min={0}
              prefixCls="rc-slider"
              reverse={false}
              step={1}
              upperBound={0}
              vertical={false}
            >
              <div
                className="rc-slider-step"
              />
            </Steps>
            <Handle
              className="rc-slider-handle"
              disabled={true}
              key="0"
              max={11}
              min={0}
              offset={0}
              prefixCls="rc-slider"
              reverse={false}
              style={Object {}}
              value={0}
              vertical={false}
            >
              <div
                aria-disabled={true}
                aria-valuemax={11}
                aria-valuemin={0}
                aria-valuenow={0}
                className="rc-slider-handle"
                onBlur={[Function]}
                onKeyDown={[Function]}
                onMouseDown={[Function]}
                role="slider"
                style={
                  Object {
                    "left": "0%",
                    "right": "auto",
                    "transform": "translateX(-50%)",
                  }
                }
                tabIndex={null}
              />
            </Handle>
            <Marks
              className="rc-slider-mark"
              included={false}
              lowerBound={0}
              marks={Object {}}
              max={11}
              min={0}
              onClickLabel={[Function]}
              reverse={false}
              upperBound={0}
              vertical={false}
            >
              <div
                className="rc-slider-mark"
              />
            </Marks>
          </div>
        </ComponentEnhancer(undefined)>
      </HourSlider>
    </div>
    <div
      className="weather__daily"
    >
      <div
        className="weather__day"
      >
        <t
          ago={false}
          calendar={false}
          decimal={false}
          element={null}
          filter={[Function]}
          format="[--]"
          fromNow={false}
          interval={60000}
          local={false}
          onChange={[Function]}
          titleFormat=""
          toNow={false}
          unit={null}
          unix={true}
          utc={false}
          withTitle={false}
        >
          <time
            dateTime={"1970-01-01T00:00:00.000Z"}
          >
            --
          </time>
        </t>
        <HourSlider
          name="default"
        >
          <img
            className="weather__icon"
            src="<svg version=\\"1.1\\" id=\\"Layer_1\\" xmlns=\\"http://www.w3.org/2000/svg\\" x=\\"0\\" y=\\"0\\" width=\\"100\\" height=\\"100\\" xml:space=\\"preserve\\"><path d=\\"M63.295 43.878h-.125l-.525 2.774-1.274 3.873h3.748l-1.301-3.898z\\"/><path d=\\"M50 20c-16.568 0-30 13.432-30 30s13.432 30 30 30c16.569 0 30-13.432 30-30S66.568 20 50 20zm-7.448 37.273h-2.499l-6.572-9.247-1.375-2.75h-.125l.399 2.75v9.122h-3.248V39.53h2.499l6.622 9.321 1.325 2.674h.125l-.4-2.674v-9.196h3.249v17.618zm4.848 3.373l-2.449-1.149 8.921-20.143 2.449 1.1L47.4 60.646zm19.868-3.498l-1.398-3.873h-5.373l-1.324 3.873h-3.424l6.297-17.62h2.523l6.299 17.62h-3.6z\\"/></svg>"
          />
        </HourSlider>
        <p
          className="weather__temp"
        >
          --
        </p>
        <p
          className="weather__temp weather__temp--low"
        >
          --
        </p>
      </div>
    </div>
    <p
      className="weather__icon-link"
    >
      <a
        href="https://www.freepik.com/free-vector/weather-icons-set_709126.htm"
        target="_blank"
      >
        Icons created by freepik - www.freepik.com
      </a>
    </p>
  </div>
</WeatherComponent>
`;
